colin/utils/cont.py

```diff
@@ -0,0 +1,275 @@
+
+        self.metadata = None
+        self._labels = None
+
+        self._pull_image()
+
+
+    @property
+    def tmpdir(self):
+        """ Temporary directory holding all the runtime data. """
+        if self._tmpdir is None:
+            self._tmpdir = mkdtemp(prefix="colin-", dir="/var/tmp")
+        return self._tmpdir
+
+    @property
+    def ostree_path(self):
+        """ ostree repository -- content """
+        if self._ostree_path is None:
+            self._ostree_path = os.path.join(self.tmpdir, "ostree-repo")
+        return self._ostree_path
+
+    @property
+    def mount_point(self):
+        """ ostree checkout -- real filesystem """
+        if self._mount_point is None:
+            self._mount_point = os.path.join(self.tmpdir, "checkout")
+            os.makedirs(self._mount_point)
+            self._checkout()
+        return self._mount_point
+
+    @property
+    def layers_path(self):
+        """ Directory with all the layers (docker save). """
+        if self._layers_path is None:
+            self._layers_path = os.path.join(self.tmpdir, "layers")
+        return self._layers_path
+
+    def _pull_image(self):
+        """ pull the image using atomic --storage ostree """
+        image_name = ImageName.parse(self.image_name)
+
+        if self.pull:
+            skopeo_source = "docker://" + image_name.name
+            atomic_source = image_name.name
+        else:
+            skopeo_source = "docker-daemon:" + image_name.name
+            atomic_source = "docker:" + image_name.name
+
+        e = os.environ.copy()
+        e["ATOMIC_OSTREE_REPO"] = self.ostree_path
+        # must not exist, ostree will create it
+        out = subprocess.check_output(
+            ["atomic", "--debug", "pull", "--storage", "ostree", atomic_source],
+            env=e)
+        logger.debug("output of atomic command:")
+        logger.debug(out)
+
+        archive_file_name = "archive.tar"
+        archive_path = os.path.join(self.tmpdir, archive_file_name)
+        skopeo_target = "docker-archive:" + archive_path
+        skopeo_cmd = ["skopeo", "copy", skopeo_source, skopeo_target]
+        out = subprocess.check_output(skopeo_cmd)
+        logger.debug("output of skopeo command:")
+        logger.debug(out)
+
+        manifest_file_name = "manifest.json"
+
+        # first extract manifest
+        extract_file_from_tarball(archive_file_name, manifest_file_name, self.tmpdir)
+        manifest_path = os.path.join(self.tmpdir, manifest_file_name)
+        with open(manifest_path) as fd:
+            j = json.load(fd)
+
+        # figure out name of the metadata file
+        metadata_file_name = j[0]["Config"]
+
+        # then extract the metadata
+        extract_file_from_tarball(archive_file_name, metadata_file_name, self.tmpdir)
+        metadata_file_path = os.path.join(self.tmpdir, metadata_file_name)
+        with open(metadata_file_path) as fd:
+            self.metadata = json.load(fd)
```

Comment:
Just refactoring nitpick:
these 4 lines could be in a method/function since they're twice here

Reply:
will do